Mathlib/Order/Monotone/Defs.lean

```diff
@@ -0,0 +1,526 @@
+/-
+Copyright (c) 2014 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Avigad, Mario Carneiro, Yaël Dillies
+-/
+import Mathlib.Logic.Function.Iterate
+import Mathlib.Order.Basic
+import Mathlib.Tactic.Coe
+import Mathlib.Util.AssertExists
+
+/-!
+# Monotonicity
+
+This file defines (strictly) monotone/antitone functions. Contrary to standard mathematical usage,
+"monotone"/"mono" here means "increasing", not "increasing or decreasing". We use "antitone"/"anti"
+to mean "decreasing".
+
+## Definitions
+
+* `Monotone f`: A function `f` between two preorders is monotone if `a ≤ b` implies `f a ≤ f b`.
+* `Antitone f`: A function `f` between two preorders is antitone if `a ≤ b` implies `f b ≤ f a`.
+* `MonotoneOn f s`: Same as `Monotone f`, but for all `a, b ∈ s`.
+* `AntitoneOn f s`: Same as `Antitone f`, but for all `a, b ∈ s`.
+* `StrictMono f` : A function `f` between two preorders is strictly monotone if `a < b` implies
+  `f a < f b`.
+* `StrictAnti f` : A function `f` between two preorders is strictly antitone if `a < b` implies
+  `f b < f a`.
+* `StrictMonoOn f s`: Same as `StrictMono f`, but for all `a, b ∈ s`.
+* `StrictAntiOn f s`: Same as `StrictAnti f`, but for all `a, b ∈ s`.
+
+## Main theorems
+
+* `monotone_nat_of_le_succ`, `monotone_int_of_le_succ`: If `f : ℕ → α` or `f : ℤ → α` and
+  `f n ≤ f (n + 1)` for all `n`, then `f` is monotone.
+* `antitone_nat_of_succ_le`, `antitone_int_of_succ_le`: If `f : ℕ → α` or `f : ℤ → α` and
+  `f (n + 1) ≤ f n` for all `n`, then `f` is antitone.
+* `strictMono_nat_of_lt_succ`, `strictMono_int_of_lt_succ`: If `f : ℕ → α` or `f : ℤ → α` and
+  `f n < f (n + 1)` for all `n`, then `f` is strictly monotone.
+* `strictAnti_nat_of_succ_lt`, `strictAnti_int_of_succ_lt`: If `f : ℕ → α` or `f : ℤ → α` and
+  `f (n + 1) < f n` for all `n`, then `f` is strictly antitone.
+
+## Implementation notes
+
+Some of these definitions used to only require `LE α` or `LT α`. The advantage of this is
+unclear and it led to slight elaboration issues. Now, everything requires `Preorder α` and seems to
+work fine. Related Zulip discussion:
+https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Order.20diamond/near/254353352.
+
+## TODO
+
+The above theorems are also true in `ℕ+`, `Fin n`... To make that work, we need `SuccOrder α`
```

Comment:
Probably any such work would go into Basic, not Defs, so the to-do should stay there